Mathlib/CategoryTheory/Subpresheaf/Image.lean

```diff
@@ -1,84 +1,183 @@
+
+variable (p : F' ⟶ F)
+
+/-- Given a morphism `p : F' ⟶ F` of presheaves of types, this is the morphism
+from `F'` to its range. -/
+def toRange :
+    F' ⟶ (range p).toPresheaf :=
+  lift p (by rfl)
 
 @[reassoc (attr := simp)]
-theorem toImagePresheaf_ι (f : F' ⟶ F) : toImagePresheaf f ≫ (imagePresheaf f).ι = f :=
-  (imagePresheaf f).lift_ι _ _
-
-theorem imagePresheaf_comp_le (f₁ : F ⟶ F') (f₂ : F' ⟶ F'') :
-    imagePresheaf (f₁ ≫ f₂) ≤ imagePresheaf f₂ := fun U _ hx => ⟨f₁.app U hx.choose, hx.choose_spec⟩
-
-instance isIso_toImagePresheaf {F F' : Cᵒᵖ ⥤ Type w} (f : F ⟶ F') [hf : Mono f] :
-  IsIso (toImagePresheaf f) := by
-  have : ∀ (X : Cᵒᵖ), IsIso ((toImagePresheaf f).app X) := by
-    intro X
-    rw [isIso_iff_bijective]
-    constructor
-    · intro x y e
-      have := (NatTrans.mono_iff_mono_app f).mp hf X
-      rw [mono_iff_injective] at this
-      exact this (congr_arg Subtype.val e :)
-    · rintro ⟨_, ⟨x, rfl⟩⟩
-      exact ⟨x, rfl⟩
-  apply NatIso.isIso_of_isIso_app
+lemma toRange_ι : toRange p ≫ (range p).ι = p := rfl
+
+lemma toRange_app_val {i : Cᵒᵖ} (x : F'.obj i) :
+    ((toRange p).app i x).val = p.app i x := by
+  simp [toRange]
+
+@[simp]
+lemma range_toRange : range (toRange p) = ⊤ := by
+  ext i ⟨x, hx⟩
+  dsimp at hx ⊢
+  simp only [Set.mem_range, Set.mem_univ, iff_true]
+  simp only [Set.mem_range] at hx
+  obtain ⟨y, rfl⟩ := hx
+  exact ⟨y, rfl⟩
+
+lemma epi_iff_range_eq_top :
+    Epi p ↔ range p = ⊤ := by
+  simp [NatTrans.epi_iff_epi_app, epi_iff_surjective, Subpresheaf.ext_iff, funext_iff,
+    Set.range_eq_univ]
+
+instance : Epi (toRange p) := by simp [epi_iff_range_eq_top]
+
+instance [Mono p] : IsIso (toRange p) := by
+  have := mono_of_mono_fac (toRange_ι p)
+  rw [NatTrans.isIso_iff_isIso_app]
+  intro i
+  rw [isIso_iff_bijective]
+  constructor
+  · rw [← mono_iff_injective]
+    infer_instance
+  · rw [← epi_iff_surjective]
+    infer_instance
+
+lemma range_comp_le (f : F ⟶ F') (g : F' ⟶ F'') :
+    range (f ≫ g) ≤ range g := fun _ _ _ ↦ by aesop
+
+end range
+
+section image
+
+variable (G : Subpresheaf F) (f : F ⟶ F')
+
+/-- The image of a subpresheaf by a morphism of presheaves of types. -/
+@[simps]
+def image : Subpresheaf F' where
+  obj i := (f.app i) '' (G.obj i)
+  map := by
+    rintro Δ Δ' φ _ ⟨x, hx, rfl⟩
+    exact ⟨F.map φ x, G.map φ hx, by apply FunctorToTypes.naturality⟩
+
+lemma image_top : (⊤ : Subpresheaf F).image f = range f := by aesop
+
+lemma image_comp (g : F' ⟶ F'') :
+    G.image (f ≫ g) = (G.image f).image g := by aesop
+
+lemma range_comp (g : F' ⟶ F'') :
+    range (f ≫ g) = (range f).image g := by aesop
+
+end image
+
+section preimage
+
+/-- The preimage of a subpresheaf by a morphism of presheaves of types. -/
+@[simps]
+def preimage (G : Subpresheaf F) (p : F' ⟶ F) : Subpresheaf F' where
+  obj n := p.app n ⁻¹' (G.obj n)
+  map f := (Set.preimage_mono (G.map f)).trans (by
+    simp only [Set.preimage_preimage, FunctorToTypes.naturality _ _ p f]
+    rfl)
+
```

Comment:
Can you also provide `preimage_comp`? Thanks!

Reply:
Thanks for the suggestion!